### PR TITLE
Don't include the previous tests fail hook output in the search for issues

### PR DESCRIPTION
**Pull request by Vogtinator**
After the main test code runs, search_for_expected_serial_failures looks
through the generated serial output for certain patterns. In the case that the
test failed, the post_fail_hook runs after that, which might also produce output
on the serial console. Before this commit, this was not skipped and so the call
to search_for_expected_serial_failures on the next module sees it.

This can even have a cascading effect if there's a OOM message in the log files,
if it was missed in an earlier module. Then the post_fail_hook would print that
to the serial console and thus cause the next module to fail, which in turn
prints the message again and the cycle repeats.

See also https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/13610
Related ticket: https://progress.opensuse.org/issues/81382

Before: http://10.160.67.86/tests/1122
After: http://10.160.67.86/tests/1127

### CODE REVIEW

File: basetest.pm [hunk 1]
```diff
@@ -320,6 +320,11 @@ sub run_post_fail {
         eval { $self->post_fail_hook; };
         bmwqemu::diag("post_fail_hook failed: $@") if $@;
         $self->{post_fail_hook_running} = 0;
+
+        # There might be more messages on serial now.
+        # Update the offset to not stumble upon them in the next module.
+        my $json = $self->get_serial_output_json($serial_file_pos);
+        $serial_file_pos = $json->{position};
```
Comment (Martchus) [inline basetest.pm:327]: It doesn't look like we're ever calling `get_serial_output_json` without this assignment. So maybe move `$serial_file_pos = $json->{position};` into that function?

**Review by Martchus — COMMENTED** [review]
I suppose that makes sense. I have only one remark and also don't approve immediately to give other reviewers the chance to have a look (before mergify would merge it automatically).

File: basetest.pm (hunk 1 above)
Comment (Vogtinator) [reply basetest.pm]: I changed `get_serial_output_json` a bit, how's the new approach? VR is updated.

Comment (Martchus) [reply basetest.pm]: Looks good.

Comment (okurz) [reply basetest.pm]: https://github.com/os-autoinst/os-autoinst/runs/4117429312?check_suite_focus=true#step:3:581 says "basetest::get_serial_output_json does not exist! at ./17-basetest.t line 69."

Comment (Vogtinator) [reply basetest.pm]: Fixed

**Review by Martchus — APPROVED** [review]

**Comment by Martchus** [comment]
It would be nice to have test coverage for the function, though.

**Comment by Vogtinator** [comment]
> It would be nice to have test coverage for the function, though.

Yep. Unfortunately that requires mocking the backend function, which would have more logic than the function being tested...

**Comment by Vogtinator** [comment]
Ping

**Comment by okurz** [comment]
In the end we still want to have better test coverage. If you are motivated to do that, this would be great, if not, we will merge without and look into adding test coverage afterwards.

**Comment by Vogtinator** [comment]
> In the end we still want to have better test coverage. If you are motivated to do that, this would be great, if not, we will merge without and look into adding test coverage afterwards.

I'm motivated, but it needs much more time (and/or perl expertise) than I have.

If you have an example which mocks a backend method with a non-trivial function, that would simplify it quite a bit.

**Comment by okurz** [comment]
I think t/17-basetest.t itself is a good example. Maybe also https://github.com/os-autoinst/os-autoinst/blob/master/t/29-backend-ipmi.t and https://github.com/os-autoinst/os-autoinst/blob/master/t/29-backend-generalhw.t . In general I prefer trivial functions over non-trivial so maybe slightly refactor to make mocking easier :)

**Comment by Vogtinator** [comment]
> I think t/17-basetest.t itself is a good example. Maybe also https://github.com/os-autoinst/os-autoinst/blob/master/t/29-backend-ipmi.t and https://github.com/os-autoinst/os-autoinst/blob/master/t/29-backend-generalhw.t . In general I prefer trivial functions over non-trivial so maybe slightly refactor to make mocking easier :)

The hard part is that to test this PR, it's necessary to mock at the `myjsonrpc` or `backend` layers. It would be easy to mock `get_new_serial_output`, but then the code of this PR wouldn't be used during the test.

**Comment by okurz** [comment]
Mocking of read_json is already done in t/03-testapi.t so maybe that helps as reference

**Comment by mloviska** [comment]
[The cascading effect](http://kepler.suse.cz/tests/8877#step/rails/31). If there are no objections, could we merge the PR as it affects all subsequent test cases ?

**Comment by Vogtinator** [comment]
> Mocking of read_json is already done in t/03-testapi.t so maybe that helps as reference

That does indeed help.

I created a new unit test specifically for `get_new_serial_output`, see the second commit. The complexity of the test code is still more than the function which it tests (and also needed more attempts to get it working!), but it does work.

**Comment by Vogtinator** [comment]
@okurz @Martchus Is it intentional that the approval is "sticky" even across major changes like this, i.e. mergify merges those even though they weren't actually reviewed?

**Comment by Martchus** [comment]
I'd say it is deemed questionable but also accepted for now.

**Pull request merged**